packages/bus/src/Brokers/RMQEventBus.ts

```diff
@@ -0,0 +1,395 @@
+  private readonly container!: DependencyContainer;
+  private readonly backOffRetry!: RetryStrategy;
+
+  private readonly REPLY_QUEUE_NAME = 'amq.rabbitmq.reply-to';
+  private readonly APP_QUEUE_NAME = 'app';
+  private readonly DEFAULT_RECONNECT_TIMES = 20;
+
+  constructor(
+    @inject(Configuration) private readonly configuration: Configuration,
+    @inject(ConnectionFactory)
+    private readonly connectionFactory: ConnectionFactory<Connection>
+  ) {
+    this.container = configuration.container;
+    this.subject.setMaxListeners(Infinity);
+    this.backOffRetry = new ExponentialBackoffRetryStrategy({
+      maxDepth: this.DEFAULT_RECONNECT_TIMES
+    });
+  }
+
+  public async init(): Promise<void> {
+    const connectTimeout = this.configuration.connectTimeout;
+    const heartbeat = this.configuration.heartbeatInterval;
+    const credentials = this.configuration.credentials;
+    const bus = this.configuration.bus;
+
+    const buildedUrl = this.buildUrl(bus, credentials, heartbeat || 5000);
+
+    this.client = await this.backOffRetry.acquire(async () =>
+      this.connectionFactory.create(buildedUrl, {
+        timeout: connectTimeout
+      })
+    );
+
+    await this.createConsumerChannel();
+  }
+
+  public async register<T, R>(
+    type: EventHandlerConstructor<T, R>
+  ): Promise<void> {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    if (!this.container.isRegistered(type)) {
+      throw new EventHandlerNotFound(type.name);
+    }
+    const handlerInstance = this.container.resolve(type);
+
+    const eventNames: string[] = this.reflectEventsNames(type);
+    if (eventNames?.length === 0) {
+      throw new NoSubscriptionsFound(handlerInstance);
+    }
+
+    for (const eventName of eventNames) {
+      const eventHandlers = this.handlers.get(eventName);
+      if (Array.isArray(eventHandlers)) {
+        eventHandlers.push(handlerInstance);
+      } else {
+        await this.channel?.bindQueue(
+          this.configuration.clientQueue,
+          this.configuration.exchange,
+          eventName
+        );
+        this.handlers.set(eventName, [handlerInstance]);
+      }
+    }
+  }
+
+  public async unregister<T, R>(
+    type: EventHandlerConstructor<T, R>
+  ): Promise<void> {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    if (!this.container.isRegistered(type)) {
+      throw new EventHandlerNotFound(type.name);
+    }
+    const handlerInstance = this.container.resolve(type);
+
+    const eventNames: string[] = this.reflectEventsNames(type);
+    if (eventNames?.length === 0) {
+      throw new NoSubscriptionsFound(handlerInstance);
+    }
+
+    for (const eventName of eventNames) {
+      const eventHandlers = this.handlers.get(eventName);
+      if (!eventHandlers) {
+        continue;
+      }
+
+      const handlerIndex = eventHandlers.indexOf(handlerInstance);
+      eventHandlers.splice(handlerIndex, 1);
```

Comment:
In case when `handlerIndex` is -1, it will remove the last item in the list
```suggestion
      if (handlerIndex !== -1) {
        eventHandlers.splice(handlerIndex, 1);
      }
```